internal/overlord/cmdstate/handlers.go

```diff
@@ -0,0 +1,517 @@
+	if e.splitStderr {
+		e.websockets[wsStderr] = nil
+	}
+
+	// Store the execution object on the manager (for Connect).
+	m.executionsMutex.Lock()
+	m.executions[task.ID()] = e
+	m.executionsMutex.Unlock()
+	m.executionsCond.Broadcast() // signal that Connects can start happening
+	defer func() {
+		m.executionsMutex.Lock()
+		delete(m.executions, task.ID())
+		m.executionsMutex.Unlock()
+	}()
+
+	// Run the command! Killing the tomb will terminate the command.
+	ctx := tomb.Context(context.Background())
+	return e.do(ctx, task)
+}
+
+var websocketUpgrader = websocket.Upgrader{
+	CheckOrigin:      func(r *http.Request) bool { return true },
+	HandshakeTimeout: handshakeTimeout,
+}
+
+func (e *execution) connect(r *http.Request, w http.ResponseWriter, id string) error {
+	e.websocketsLock.Lock()
+	conn, ok := e.websockets[id]
+	e.websocketsLock.Unlock()
+	if !ok {
+		return os.ErrNotExist
+	}
+	if conn != nil {
+		return fmt.Errorf("websocket %q already connected", id)
+	}
+
+	// Upgrade the HTTP connection to a websocket connection.
+	conn, err := websocketUpgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return err
+	}
+
+	// Save the connection.
+	e.websocketsLock.Lock()
+	defer e.websocketsLock.Unlock()
+	e.websockets[id] = conn
+
+	// Signal that we're connected.
+	if id == wsControl {
+		close(e.controlConnected)
+	} else if e.websockets[wsStdio] != nil && (!e.splitStderr || e.websockets[wsStderr] != nil) {
+		close(e.ioConnected)
+	}
+	return nil
+}
+
+func (e *execution) getWebsocket(key string) *websocket.Conn {
+	e.websocketsLock.Lock()
+	defer e.websocketsLock.Unlock()
+	return e.websockets[key]
+}
+
+// waitIOConnected waits till all the I/O websockets are connected or the
+// connect timeout elapses (or the provided ctx is cancelled).
+func (e *execution) waitIOConnected(ctx context.Context, execID string) error {
+	ctx, cancel := context.WithTimeout(ctx, connectTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			logger.Noticef("Exec %s: timeout waiting for websocket connections: %v", execID, ctx.Err())
+			return fmt.Errorf("exec %s: timeout waiting for websocket connections: %w", execID, ctx.Err())
+		}
+		return ctx.Err()
+	case <-e.ioConnected:
+		return nil
+	}
+}
+
+// do actually runs the command.
+func (e *execution) do(ctx context.Context, task *state.Task) error {
+	// Wait till client has connected to "stdio" websocket (and "stderr" if
+	// separating stderr), to avoid race conditions forwarding I/O.
+	err := e.waitIOConnected(ctx, task.ID())
+	if err != nil {
+		return err
+	}
+
+	// Files/pipes to close before and after waiting for output to be finished sending.
+	var beforeClosers []io.Closer
+	var afterClosers []io.Closer
+
+	// Stdin/stdout/stderr for the exec.Cmd process.
+	var stdin *os.File
+	var stdout *os.File
+	var stderr *os.File
+
+	// Closed to make the controlLoop exit early.
+	controlExit := make(chan struct{})
+	defer close(controlExit)
```

Comment:
Related to earlier control. This should end up with the same name.

Reply:
Done: all `stopControl` now, which is significantly clearer.